src/table/io/OutputArchive.cc

```diff
@@ -99,6 +100,15 @@ class OutputArchive::Impl {
         } catch (pex::exceptions::NotFoundError &) {
             iter->getTable()->getMetadata()->add("AR_NAME", name, "Class name for objects stored here");
         }
+        // Also add it as the EXTNAME.
+        try {
```

Comment:
Are we doing this sort of "overwrite existing else add new one" logic in other places? I'm a bit confused because I thought the PropertyList `set()` method did this already. The `add()` method is designed for "add a new keyword here in addition to what's already there" but `set()` is "replace anything there already with this new one". So why isn't this code using `set()`? (I understand this is copying the implementation above).

Reply:
Yeah, I was just stupidly copying the above. I changed it to  `set` and it works fine; do you want me to do the same change for AR_NAME?

Reply:
Yes, please change AR_NAME as well -- it will continue to confuse people I think. If it is like that for some reason we can't comprehend I'd hope someone would put a more explicit comment in there. `set()` is used elsewhere in the file.

Reply:
`git blame` indicates that @TallJimbo wrote the code so maybe there's some subtlety here we are not understanding.

Reply:
No idea, I'm afraid.  I agree it looks like it should just be a `set`.

Reply:
Using `set` (for AR_NAME) causes [this test](https://github.com/lsst/afw/blob/4939cbb3d9e4416b6dd5c2f40f6cf6e070342a3c/tests/test_tableArchives.cc#L714-L716) to fail:

```_____________________________________________________________ UtilsBinaryTester.test_exe__project_mschmitz_repos_afw_tests_test_tableArchives _____________________________________________________________

args = (<test_executables.UtilsBinaryTester testMethod=test_exe__project_mschmitz_repos_afw_tests_test_tableArchives>,)
self = <test_executables.UtilsBinaryTester testMethod=test_exe__project_mschmitz_repos_afw_tests_test_tableArchives>

    def test_executable_runs(*args):
        self = args[0]
>       self.assertExecutable(executable)

/software/lsstsw/stack_20210415/stack/miniconda3-py38_4.9.2-0.5.0/Linux64/utils/21.0.0-6-g0bf7090+18535a8d22/python/lsst/utils/tests.py:246: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/software/lsstsw/stack_20210415/stack/miniconda3-py38_4.9.2-0.5.0/Linux64/utils/21.0.0-6-g0bf7090+18535a8d22/python/lsst/utils/tests.py:218: in assertExecutable
    self.fail(msg)
E   AssertionError: Bad exit status from '/project/mschmitz/repos/afw/tests/test_tableArchives': 201
------------------------------------------------------------------------------------------ Captured stdout call -------------------------------------------------------------------------------------------
Running executable '/project/mschmitz/repos/afw/tests/test_tableArchives' with no arguments...
Running 15 test cases...
afw.image.fits.ExposureFitsReader WARN: Could not read PSF; setting to null: PersistableFactory with name 'DummyPsf' not found, and import of module 'testTableArchivesLib' failed (possibly because Python calls were not available from C++). {0}; loading object with id=1, name='DummyPsf' {1}
tests/test_tableArchives.cc(714): error: in "ArchiveMetadata": check names.size() == 2u has failed [1 != 2]
tests/test_tableArchives.cc(715): error: in "ArchiveMetadata": check names[0] == "Footprint" has failed [HeavyFootprintF != Footprint]
unknown location(0): fatal error: in "ArchiveMetadata": memory access violation at address: 0x00000030: no mapping at fault address
tests/test_tableArchives.cc(716): last checkpoint
```

Reply:
Memory access violation is really interesting. If you run the test outside pytest do you get a stack trace? The code assumes that getMetadata returns something so it's blowing my mind a bit that this is somehow dying in `set`.

Reply:
Um... How do I run a  `.cc` test on its own?

Reply:
I mean run "python tests/test_file_that_fails.py"

Reply:
Hmmmm! No error:

```.Running executable '/project/mschmitz/repos/afw/tests/test_tableArchives' with no arguments...
--------------------------------------------------------------------------------------
The following warning is expected, and is an indication the test has passed.
--------------------------------------------------------------------------------------

*** No errors detected
Running 15 test cases...
afw.image.fits.ExposureFitsReader WARN: Could not read PSF; setting to null: PersistableFactory with name 'DummyPsf' not found, and import of module 'testTableArchivesLib' failed (possibly because Python calls were not available from C++). {0}; loading object with id=1, name='DummyPsf' {1}
```

Reply:
are you running it from a built afw? That library should be loadable.

Reply:
Yes, locally rebuilt afw on top of the shared stack on lsst-devl

Reply:
I'll see if I can do a local build of this branch.

Reply:
Thanks! I repushed with the change (commit 17610c8)

Reply:
Okay. Maybe I just built the wrong version then
```
$ rebuild -r tickets/DM-28716 afw
Fetching Products...
Fetching afw...
                                            afw ok [tickets/DM-28716] (2.2 sec).
Fetching daf_base...
Fetching daf_persistence...
Fetching pex_config...
Fetching geom...
Fetching log...
Fetching sphgeom...
Fetching utils...
Fetching astshim...
                                               pex_config ok [master] (1.2 sec).
Fetching afwdata...
                                                      log ok [master] (1.3 sec).
Fetching base...
                                                  sphgeom ok [master] (1.3 sec).
Fetching sconsUtils...
                                                  astshim ok [master] (1.3 sec).
                                                 daf_base ok [master] (1.5 sec).
Fetching pex_exceptions...
                                                     geom ok [master] (1.5 sec).
                                                    utils ok [master] (1.5 sec).
                                          daf_persistence ok [master] (1.6 sec).
                                           pex_exceptions ok [master] (0.9 sec).
                                                     base ok [master] (1.1 sec).
                                               sconsUtils ok [master] (1.1 sec).
                                                  afwdata ok [master] (1.2 sec).
Performing git-lfs pulls...
Pulling afwdata LFS data...
                                                           afwdata ok (0.1 sec).
# BUILD ID: b5356
Saving environment information in /Volumes/MediaHD/Users/timj/work/lsstsw/build/builds/b5356.env
             afwdata: 21.0.0-2-g294ccec (already installed).
          sconsUtils: 21.0.0-11-ge0be345 (already installed).
             astshim: 21.0.0-2-g45278ab+64c1bc5aa5 (already installed).
                base: 21.0.0-7-gdf92d54+64c1bc5aa5 (already installed).
             sphgeom: 21.0.0-1-g8760c09+64c1bc5aa5 (already installed).
      pex_exceptions: 21.0.0-2-gde069b7+4f46bdaea8 (already installed).
               utils: 21.0.0-6-g0bf7090+18535a8d22 (already installed).
            daf_base: 21.0.0-3-gd7ab7aa+fdc5edd43f (already installed).
                geom: 21.0.0-5-g8c1d971+7b9a448d34 (already installed).
                 log: 21.0.0-5-gcc89fd6+fdc5edd43f (already installed).
     daf_persistence: 21.0.0-8-g5674e7b+8087abed30 (already installed).
          pex_config: 21.0.0-1-ga51b5d4+5491f2a448 (already installed).
                 afw: tickets.DM-28716-g1a1c2439e9+fb0e51fe34 .................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................ok (1425.4 sec).
# BUILD b5356 completed.
```

Reply:
I've finally read the error message and got it to fail. I thought it was a segv but it's actually a simple C++ test failure.

```
./test_tableArchives 
Running 15 test cases...
--------------------------------------------------------------------------------------
The following warning is expected, and is an indication the test has passed.
--------------------------------------------------------------------------------------
afw.image.fits.ExposureFitsReader WARN: Could not read PSF; setting to null: PersistableFactory with name 'DummyPsf' not found, and import of module 'testTableArchivesLib' failed (possibly because Python calls were not available from C++). {0}; loading object with id=1, name='DummyPsf' {1}
tests/test_tableArchives.cc:714: error: in "ArchiveMetadata": check names.size() == 2u has failed [1 != 2]
tests/test_tableArchives.cc:715: error: in "ArchiveMetadata": check names[0] == "Footprint" has failed [HeavyFootprintF != Footprint]
tests/test_tableArchives.cc:716: error: in "ArchiveMetadata": check names[1] == "HeavyFootprintF" has failed [ != HeavyFootprintF]

*** 3 failures are detected in the test module "table - archives"
```
and these tests are reading AR_NAME header I guess? @TallJimbo is the DummyPsf warning a red herring? I think I need to read the test because it naively looks like it thought it was getting two headers for the name but is actually only getting one and it says HeavyFootprintF.

Reply:
```c++
    auto names = metadata.getArray<std::string>("AR_NAME");
    std::sort(names.begin(), names.end());
    BOOST_CHECK_EQUAL(names.size(), 2u);
    BOOST_CHECK_EQUAL(names[0], "Footprint");
    BOOST_CHECK_EQUAL(names[1], "HeavyFootprintF");
```
so I don't understand how this test is consistent with the code we removed that explicitly stated that it wanted to ensure only one AR_NAME.

Reply:
I think we misunderstood the original motivation for the code we replaced for AR_NAME. I think we thought that it was trying to ensure only one AR_NAME is used. What it's actually doing is let you set multiple distinct AR_NAME values but ensures that it doesn't end up with setting a value that is already set. I guess we need to put the AR_NAME code back but leave the EXTNAME code using set and state that the most recent AR_NAME given is the value we want to use.

Reply:
Ah, sorry, I should have remembered this.  In retrospect it's probably too clever, but the system can save objects of different types to the same HDU if they have the same on-disk representation, and that does happen once in a while.